crates/symbolicator/src/services/symbolication/module_lookup.rs

```diff
@@ -51,23 +50,142 @@ impl<'a> SymCacheLookupResult<'a> {
 
 pub struct SourceObject(SelfCell<ByteView<'static>, Object<'static>>);
 
-struct SourceObjectEntry {
+struct ModuleEntry {
     module_index: usize,
     object_info: CompleteObjectInfo,
+    symcache: Option<Arc<SymCacheFile>>,
     source_object: Option<SourceObject>,
 }
 
-pub struct SourceLookup {
-    inner: Vec<SourceObjectEntry>,
+pub struct ModuleLookup {
+    modules: Vec<ModuleEntry>,
+    scope: Scope,
+    sources: Arc<[SourceConfig]>,
 }
 
-impl SourceLookup {
+impl ModuleLookup {
+    /// Creates a new [`ModuleLookup`] out of the given module iterator.
+    pub fn new<Iter>(scope: Scope, sources: Arc<[SourceConfig]>, iter: Iter) -> Self
+    where
+        Iter: IntoIterator<Item = CompleteObjectInfo>,
+    {
+        let mut modules: Vec<_> = iter
+            .into_iter()
+            .enumerate()
+            .map(|(module_index, object_info)| ModuleEntry {
+                module_index,
+                object_info,
+                symcache: None,
+                source_object: None,
+            })
+            .collect();
+
+        modules.sort_by_key(|entry| entry.object_info.raw.image_addr.0);
+
+        // Ignore the name `dedup_by`, I just want to iterate over consecutive items and update
+        // some.
```

Comment:
#686 adds a few more comments explaining the weirdness there. The thing that makes it a bit weird is the borrow checker, since you can’t have two mut borrows to elements of the slice